cpp/Ice/filesystem/Memory.cpp

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) ZeroC, Inc.
+
+#include "Memory.h"
+
+using namespace std;
+using namespace Memory;
+
+//
+// MNode
+//
+
+MNode::MNode(string name, MDirectoryPtr parent) : _name(std::move(name)), _parent(std::move(parent))
+{
+    // Create an identity. The root directory has the fixed identity "RootDir"
+    _id.name = _parent ? Ice::generateUUID() : "RootDir";
+}
+
+string
+MNode::name(const Ice::Current&)
+{
+    return _name;
+}
+
+void
+MNode::activate(const Ice::ObjectAdapterPtr& adapter)
+{
+    auto proxy = adapter->add<Filesystem::NodePrx>(shared_from_this(), _id);
+    if (_parent)
+    {
+        _parent->addChild(proxy);
+    }
+}
```

Comment:
_parent is only used for the name and and adding a child. I think we should just do all of this in the server setup. 

1. activate doesn't "activate" anything.
2. `if(_parent)` is true 99% of the time
3. we could remove our dependance on shared_from_this.

Reply:
It's only 3 extra lines of code. For the 3 things we call activate on that aren't the root node.

Reply:
Simplified as suggested.